examples/swap-tracks/main.go

```diff
@@ -0,0 +1,196 @@
+		webrtc.RtpTransceiverInit{Direction: webrtc.RTPTransceiverDirectionRecvonly})
+	if err != nil {
+		panic(err)
+	}
+	_, err = peerConnection.AddTransceiverFromKind(webrtc.RTPCodecTypeVideo,
+		webrtc.RtpTransceiverInit{Direction: webrtc.RTPTransceiverDirectionRecvonly})
+	if err != nil {
+		panic(err)
+	}
+
+	// Set the remote SessionDescription
+	err = peerConnection.SetRemoteDescription(offer)
+	if err != nil {
+		panic(err)
+	}
+
+	// Which track is currently being handled
+	currTrack := 0
+	// The total number of tracks
+	trackCount := 0
+	// The channel of packets with a bit of buffer
+	packets := make(chan *rtp.Packet, 60)
+
+	// Set a handler for when a new remote track starts
+	peerConnection.OnTrack(func(track *webrtc.Track, receiver *webrtc.RTPReceiver) {
+		fmt.Printf("Track has started, of type %d: %s \n", track.PayloadType(), track.Codec().Name)
+		trackNum := trackCount
+		trackCount++
+		// The last timestamp so that we can change the packet to only be the delta
+		var lastTimestamp uint32
+		// Whether this track is the one currently sending to the channel (on change
+		// of this we send a PLI to have the entire picture updated)
+		var onMe bool
+		for {
+			// Read RTP packets being sent to Pion
+			rtp, readErr := track.ReadRTP()
+			if readErr != nil {
+				panic(readErr)
+			}
+
+			// Change the timestamp to only be the delta
+			oldTimestamp := rtp.Timestamp
+			if lastTimestamp == 0 {
+				rtp.Timestamp = 0
+			} else {
+				rtp.Timestamp -= lastTimestamp
+			}
+			lastTimestamp = oldTimestamp
+
+			// Check if this is the current track
+			if currTrack == trackNum {
+				// If just switched to this track, send PLI to get picture refresh
+				if !onMe {
+					onMe = true
+					writeErr := peerConnection.WriteRTCP([]rtcp.Packet{&rtcp.PictureLossIndication{MediaSSRC: track.SSRC()}})
+					if writeErr != nil {
+						fmt.Println(writeErr)
+					}
+				}
+				packets <- rtp
+			} else {
+				onMe = false
+			}
+		}
+	})
+	// Set the handler for ICE connection state and update chan if connected
+	connected := make(chan struct{}, 1)
+	peerConnection.OnICEConnectionStateChange(func(connectionState webrtc.ICEConnectionState) {
+		fmt.Printf("Connection State has changed %s \n", connectionState.String())
+		// Non-blocking just in case already sent
+		if connectionState == webrtc.ICEConnectionStateConnected {
+			select {
+			case connected <- struct{}{}:
+			default:
+			}
+		}
+	})
+
+	// Create an answer
+	answer, err := peerConnection.CreateAnswer(nil)
+	if err != nil {
+		panic(err)
+	}
+
+	// Sets the LocalDescription, and starts our UDP listeners
+	err = peerConnection.SetLocalDescription(answer)
+	if err != nil {
+		panic(err)
+	}
+
+	// Output the answer in base64 so we can paste it in browser
+	fmt.Printf("Paste below base64 in browser:\n%v\n", signal.Encode(answer))
+
+	// Asynchronously take all packets in the channel and write them out to our
+	// track
+	go func() {
```

Comment:
What do you think of dropping the channel communication and doing the input/output in the OnTrack read loop? The indirection might confuse people new to Go (who don't know channels yet)

Conceptually I think I see two code blocks

* Firing something every 5 seconds 
* Reading the packets, updating the SSRC/SequenceNumber and pushing

Reply:
This is what I had originally, but keeping the shared timestamp and sequence numbers intact w/ atomic updates from either side became a bit confusing and not as guaranteed to keep packets (or I'd have to lock the entire loop iteration). But I can change it back to that if I must,.

In this use case, I think the benefit using the channel to combine multiple sources outweighs the cost of somehow a Go dev not knowing about them.